File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/VoiceProfile/VoiceProfileViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for Voice Profile - multi-metric batch analysis via Tessera.analyze().
///
/// ## VoxaTrace Integration
/// ```swift
/// // 1. Decode audio
/// let audioData = SonixDecoder.decode(path: filePath)
///
/// // 2. Extract pitch contour
/// let extractor = PitchDetection.createContourExtractor()
/// let contour = extractor.extract(audio: samples, sampleRate: sampleRate)
/// extractor.release()
///
/// // 3. Batch analysis
/// let result = Tessera.analyze(contour: contour)
/// // result.breath, result.agility, result.vocalRange
/// ```
@MainActor
final class VoiceProfileViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var breathCapacity: Float = 0
    @Published private(set) var breathControl: Float = 0
    @Published private(set) var agilityScore: Float = 0
    @Published private(set) var vocalRangeLow: String = ""
    @Published private(set) var vocalRangeHigh: String = ""
    @Published private(set) var vocalRangeOctaves: Float = 0
    @Published private(set) var isAnalyzing = false

    // MARK: - Actions

    func analyzeOffline() {
        isAnalyzing = true
        breathCapacity = 0
        breathControl = 0
        agilityScore = 0
        vocalRangeLow = ""
        vocalRangeHigh = ""
        vocalRangeOctaves = 0

        Task {
            guard let audioURL = Bundle.main.url(forResource: "Alankaar 01_voice", withExtension: "m4a"),
                  let audioData = SonixDecoder.decode(path: audioURL.path) else {
                await MainActor.run { isAnalyzing = false }
                return
            }

            // Extract pitch contour
            let extractor = PitchDetection.createContourExtractor()
            let contour = extractor.extract(audio: audioData.samples, sampleRate: audioData.sampleRate)
            extractor.release()

            // Batch analysis via Tessera.analyze()
            let result = Tessera.analyze(contour: contour)

            var cap: Float = 0
            var ctrl: Float = 0
            var agil: Float = 0
            var low = ""
            var high = ""
            var octaves: Float = 0

            if let breath = result.breath {
                cap = breath.capacity
                ctrl = breath.controlScore
            }
            if let agility = result.agility {
                agil = agility.scores.first?.floatValue ?? 0
            }
            if let vocalRange = result.vocalRange {
                low = vocalRange.range.lower.noteLabel
                high = vocalRange.range.upper.noteLabel
                octaves = vocalRange.range.octaves
            }

            await MainActor.run {
                breathCapacity = cap
                breathControl = ctrl
                agilityScore = agil
                vocalRangeLow = low
                vocalRangeHigh = high
                vocalRangeOctaves = octaves
                isAnalyzing = false
            }
        }
    }
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/voiceprofile/viewmodel/VoiceProfileViewModel.kt
package com.musicmuni.voxatrace.demo.sections.voiceprofile.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.sonix.SonixDecoder
import com.musicmuni.voxatrace.tessera.Tessera
import com.musicmuni.voxatrace.tessera.model.TesseraMetric
import com.musicmuni.voxatrace.tona.PitchDetection
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for Voice Profile - multi-metric batch analysis via Tessera.analyze().
 *
 * APIs demonstrated:
 * - Tessera.analyze() for batch analysis (breath, agility, vocal range)
 * - PitchDetection.createContourExtractor() for pitch extraction
 * - SonixDecoder.decode() for audio decoding
 */
class VoiceProfileViewModel : ViewModel() {

    // Published state
    private val _breathCapacity = MutableStateFlow(0f)
    val breathCapacity: StateFlow<Float> = _breathCapacity.asStateFlow()

    private val _breathControl = MutableStateFlow(0f)
    val breathControl: StateFlow<Float> = _breathControl.asStateFlow()

    private val _agilityScore = MutableStateFlow(0f)
    val agilityScore: StateFlow<Float> = _agilityScore.asStateFlow()

    private val _vocalRangeLow = MutableStateFlow("")
    val vocalRangeLow: StateFlow<String> = _vocalRangeLow.asStateFlow()

    private val _vocalRangeHigh = MutableStateFlow("")
    val vocalRangeHigh: StateFlow<String> = _vocalRangeHigh.asStateFlow()

    private val _vocalRangeOctaves = MutableStateFlow(0f)
    val vocalRangeOctaves: StateFlow<Float> = _vocalRangeOctaves.asStateFlow()

    private val _isAnalyzing = MutableStateFlow(false)
    val isAnalyzing: StateFlow<Boolean> = _isAnalyzing.asStateFlow()

    // Actions

    fun analyzeOffline(context: Context) {
        viewModelScope.launch {
            _isAnalyzing.value = true
            _breathCapacity.value = 0f
            _breathControl.value = 0f
            _agilityScore.value = 0f
            _vocalRangeLow.value = ""
            _vocalRangeHigh.value = ""
            _vocalRangeOctaves.value = 0f

            try {
                val audioFile = withContext(Dispatchers.IO) {
                    copyAssetToFile(context, "Alankaar 01_voice.m4a")
                }

                val audioData = withContext(Dispatchers.IO) {
                    SonixDecoder.decode(audioFile.absolutePath)
                }

                if (audioData == null) {
                    Napier.e("Failed to decode audio file")
                    _isAnalyzing.value = false
                    return@launch
                }

                // Extract pitch contour via tona
                val extractor = PitchDetection.createContourExtractor()
                val contour = withContext(Dispatchers.IO) {
                    extractor.extract(audioData.samples, audioData.sampleRate)
                }
                extractor.release()

                // Batch analysis via Tessera.analyze()
                val result = withContext(Dispatchers.IO) {
                    Tessera.analyze(contour)
                }

                result.breath?.let { breath ->
                    _breathCapacity.value = breath.capacity
                    _breathControl.value = breath.controlScore
                }

                result.agility?.let { agility ->
                    _agilityScore.value = agility.scores.firstOrNull() ?: 0f
                }

                result.vocalRange?.let { vocalRange ->
                    _vocalRangeLow.value = vocalRange.range.lower.noteLabel
                    _vocalRangeHigh.value = vocalRange.range.upper.noteLabel
                    _vocalRangeOctaves.value = vocalRange.range.octaves
                }

            } catch (e: Exception) {
                Napier.e("Voice profile analysis failed", e)
            } finally {
                _isAnalyzing.value = false
            }
        }
    }

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }
}
